### PR TITLE
Track flaky tests (redux)

### PR DESCRIPTION
**Pull request by stephanos**
## What changed?

Uploading test results to Buildkite for analytical purposes.

⚠️ This will not work for any PR from a fork (Temporal developer or not), due to GitHub's security model. It has to be a branch on this repo to run successfully (includes `main`, ofc).

## Why?

To be able to analyze test flakiness.

## How did you test it?

Ran this PR and saw test run in Buildkite.

Screenshot 2024-05-31 at 5 11 07 PM

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -446,6 +446,22 @@ coverage-report: $(SUMMARY_COVER_PROFILE)
 	@printf $(COLOR) "Generate HTML report from $(SUMMARY_COVER_PROFILE) to $(SUMMARY_COVER_PROFILE).html..."
 	@go tool cover -html=$(SUMMARY_COVER_PROFILE) -o $(SUMMARY_COVER_PROFILE).html
 
+upload-test-results:
+	$(foreach FILE,$(wildcard $(TEST_OUTPUT_ROOT)/*.junit.xml),\
```
Comment (dnr) [inline Makefile:450]: I think you want a for loop in shell, not a make foreach construct. actually I'd want to move this all into a script in the repo and just call it here, that'll be easier to maintain

Comment (stephanos) [reply Makefile:450]: > I think you want a for loop in shell, not a make foreach construct.

Why's that? There are two other make targets here already that I was mimicking.

> actually I'd want to move this all into a script in the repo and just call it here, that'll be easier to maintain

Oh? What would make it easier to maintain? I don't believe we have any dev scripts anymore. We actually got rid of one in the past (see https://github.com/temporalio/temporal/pull/4938/files#diff-d22690e5ffc7fbcc6c0d71f7ec6914992106d70a0f0d83a09e7f665056059f53L315).

**Comment by rodrigozhou** [comment]
What is buildkite doing? Is there any way to replicate it with GHA?

**Comment by stephanos** [comment]
@rodrigozhou Buildkite is giving us test analytics, ie test performance and test flakiness. I don't think GHA has this feature.

Screenshot 2024-06-04 at 10 12 41 AM

File: Makefile (hunk 1 above)
Comment (stephanos) [reply Makefile:450]: That's a fair point. I think there's a simplicity to not invoking separate scripts that I really appreciate; but - to your point - it comes at a cost. Alright, I'll move it to a separate shell script.

Comment (stephanos) [reply Makefile]: @dnr It's a separate shell script now.

File: develop/upload-test-results.sh [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+      -H "Authorization: Token token=$BUILDKITE_ANALYTICS_TOKEN" \
+      -F "data=@$FILE" \
+      -F "format=junit" \
+      -F "run_env[CI]=github_actions" \
+      -F "run_env[key]=$GITHUB_ACTION-$GITHUB_RUN_NUMBER-$GITHUB_RUN_ATTEMPT" \
+      -F "run_env[number]=$GITHUB_RUN_NUMBER" \
+      -F "run_env[branch]=$GITHUB_REF" \
+      -F "run_env[commit_sha]=$GITHUB_SHA" \
+      -F "run_env[url]=https://github.com/$GITHUB_REPOSITORY/actions/runs/$GITHUB_RUN_ID" \
+      https://analytics-api.buildkite.com/v1/uploads || true
```
Comment (dnr) [inline develop/upload-test-results.sh:18]: no need for `|| true`, if you haven't `set -e` then failures don't do anything

File: develop/upload-test-results.sh [hunk 3]
```diff
@@ -0,0 +1,19 @@
+#!/bin/sh
+
+echo "uploading test results from $(pwd)"
+
+for FILE in *.junit.xml; do
```
Comment (dnr) [inline develop/upload-test-results.sh:5]: I'd use a lowercase name for a shell variable (the convention is environment variables are uppercase but this isn't an environment variable)

File: develop/upload-test-results.sh [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    echo "uploading $FILE"
+
+    curl -i -X POST \
+      -H "Authorization: Token token=$BUILDKITE_ANALYTICS_TOKEN" \
```
Comment (dnr) [inline develop/upload-test-results.sh:9]: I'm a little surprised shellcheck doesn't complain that the variable reference should use {}. it's fine though

**Review by dnr — COMMENTED** [review]
my other question is, are we paying for buildkite? for just this feature?

**Comment by stephanos** [comment]
@dnr I just looked and I can't find any specific pricing for this particular feature anywhere, it might be a loss leader. Whether we pay for Buildkite I'm also not sure, I see more than 1 project that actually uses it. So I think we are.

File: develop/upload-test-results.sh (hunk 3 above)
Comment (stephanos) [reply develop/upload-test-results.sh:5]: Good point; fixed!

File: develop/upload-test-results.sh (hunk 2 above)
Comment (stephanos) [reply develop/upload-test-results.sh:18]: True!

**Review by dnr — APPROVED** [review]

**Pull request merged**